metadata/com.github.luben/zstd-jni/1.5.2-5/resource-config.json

```diff
@@ -0,0 +1,13 @@
+{
+  "bundles": [],
+  "resources": {
+    "includes": [
+      {
+        "condition": {
+          "typeReachable": "com.github.luben.zstd.util.Native"
+        },
+        "pattern": "\\Qlinux/amd64/libzstd-jni-1.5.2-5.so\\E"
```

Comment:
Is there a way to add support for other operating systems here? The way I see it, it will only match linux/amd64.

Reply:
> Is there a way to add support for other operating systems here? The way I see it, it will only match linux/amd64.
- It's very simple. For my branch, you only need to execute the following commands on Windows and MacOS in the development environment where GraalVM has been configured. Of course, GraalVM does not currently support IBM Power and BSD, and it has no distributions for those two platforms.
```shell
cd ./tests/src/com.github.luben/zstd-jni/1.5.2-5/
./gradlew -Pagent clean test
./gradlew metadataCopy --task test
```
- After executing the above command, execute `./gradlew -Pagent nativeTest` at the same time to verify whether the unit test can run successfully under the GraalVM Native Image. All GraalVM reachability is generated from unit tests, so I can't be sure that the file paths in my unit tests will work correctly under Windows. The execution platform can be limited by Junit annotations.
- Refer to https://github.com/oracle/graalvm-reachability-metadata/pull/121#discussion_r1028445809 . I am generating GraalVM reachability metadata in WSL under Windows 11. I have neither Windows native nor MacOS configured. It would be great if there are friends who are willing to submit related PRs in the future.

> Also, is it possible to make it work for multiple versions? I don't expect the reflection data to be changing.

- My PR has a file with the path `metadata/com.github.luben/zstd-jni/index.json`, and the `latest` tag corresponding to `1.5.2-5` is set to `true`, which means Once the user's version does not match `1.5.2-5`, the `1.5.2-5` version of the GraalVM reachability metadata will be used. If there is a future version that needs to modify the metadata, we can open a PR for the new version.

Reply:
Thanks for explaining the versioning - I didn't know it will be applied to later versions.

Regarding the platforms support, the `pattern` is a regex, so we should be able to match all platform specific artifacts, e.g. we can use:
```
  "pattern": "[a-z]+/[a-z-0-9]+/libzstd-jni-[0-9.-]+(?:so|dll|dylyb)"
```
and then this metadata will apply for all combinations of OS / ISA.

Reply:
@luben

- Oh, I misunderstood you. I thought you meant something else. I hope to actively generate this part according to GraalVM Agent. Adding by hand seemed strange, mainly because I didn't know how to write unit tests that fit this rule.

Reply:
- In fact, for GraalVM Native Image, we only need to consider linux/amd64, windows/amd64, macos/amd64, linux/aarch64, macos/aarch64 five platforms. I think this problem needs some real testing. Considering that the github action only has amd64 machines, this is a difficult problem to solve.

Reply:
I don't think if we make it more generic it will be wrong. I think the versioning does not apply here - we are providing a pattern for the resource name - it could be versioned or not, we chose to make it versioned, but that has nothing to do with the versions used in the metadata.

Reply:
- But I really don't know what to do with it. At least after I modified it to the following content, `./gradlew -Pagent nativeTest` failed.
```shell
{
   "resources": {
   "includes": [{
       "condition":{"typeReachable":"com.github.luben.zstd.util.Native"},
       "pattern": "\\Q[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?:so|dll|dylyb)\\E"
     }]},
   "bundles": []
}
```

- ![image](https://user-images.githubusercontent.com/20187731/203318187-3e03edbe-bb64-4fd4-858a-1b716d658fa6.png)
- The error message is as follows. Conditional Agent Mode is less general. Conditional Agent Mode is a mandatory requirement for the central repository of GraalVM Reachablity Metadata.
```shell
Error: Resource pattern "\Q[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?:so|dll|dylyb)\E"s specifies unknown module \ Q[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?
```

Reply:
- After changing to the following style, `./gradlew -Pagent nativeTest` still failed. GraalVM is not smart enough.
```shell
{
   "resources": {
   "includes": [{
       "condition":{"typeReachable":"com.github.luben.zstd.util.Native"},
       "pattern":"[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?:so|dll|dylyb)"
     }]},
   "bundles": []
}
```
- ![image](https://user-images.githubusercontent.com/20187731/203323560-dc1791d2-d06a-42b7-a6d9-573ebae480b5.png)
```shell
Error: Resource pattern "[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?:so|dll|dylyb)"s specifies unknown module [a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(?
Error: Use -H:+ReportExceptionStackTraces to print stacktrace of underlying exception

```

Reply:
This passes for me:
```
"pattern": "[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(so|dll|dylyb)"
```

I guess the non-capturing marker (':?') was confusing it.

BTW, some JSON format checks were not passing for tests/src/com.github.luben/zstd-jni/1.5.2-5/user-code-filter.json, here is a patch:

```diff
diff --git a/tests/src/com.github.luben/zstd-jni/1.5.2-5/user-code-filter.json b/tests/src/com.github.luben/zstd-jni/1.5.2-5/user-code-filter.json
index f78ce06..669d71b 100644
--- a/tests/src/com.github.luben/zstd-jni/1.5.2-5/user-code-filter.json
+++ b/tests/src/com.github.luben/zstd-jni/1.5.2-5/user-code-filter.json
@@ -1,6 +1,10 @@
 {
   "rules": [
-    {"excludeClasses": "**"},
-    {"includeClasses": "com.github.luben.zstd.**"}
+    {
+      "excludeClasses": "**"
+    },
+    {
+      "includeClasses": "com.github.luben.zstd.**"
+    }
   ]
-}
\ No newline at end of file
+}
```

Reply:
- Oh oh, I forgot `./gradlew :spotlessApply`. I submitted a new commit and confirmed that `./gradlew clean test -Pcoordinates=com.github.luben:zstd-jni:1.5.2-5` in the main directory is normal. `"pattern": "[a-z]+/[a-z0-9]+/libzstd-jni-[0-9.-]+(so|dll|dylyb)"` successfully passed my local test, I think there should be no problem?
- ![image](https://user-images.githubusercontent.com/20187731/203333925-4228729b-bdd3-4cb4-ae80-3a456275a358.png)

Reply:
Thanks!